### PR TITLE
Fix regression in PR #39

### PR DESCRIPTION
**Pull request by mvastola**
PR #39 (merged in bca5db9) includes a regression wherein observers are no longer automatically instantiated on application startup. This PR seeks to:
- Fix this bug (completed)
- Add a CI test to prevent this bug from recurring (completed)

Thanks to @jachenry for initially pointing out this bug.

### CODE REVIEW

**Comment by mvastola** [comment]
@jachenry Please try pointing to master on `Partyista/rails-observers` and confirm the error goes away.

**Maintainer(s)**, this should be ready to merge as the new test I added (as well as all of the old ones) are passing. Also, a release was promised before I became aware of this bug, so I dropped that pursuit to make sure Rails 5 compatibility was working prior to a release enabling it. Now that I'm content that it is working, I'd like to re-request that this functionality be released to RubyGems after this PR is merged.

**Comment by mvastola** [comment]
@jachenry any news?

**Comment by jachenry** [comment]
@mvastola I'll try to take a look at this tomorrow and let you know what I find. Thanks for helping maintain the project.

**Comment by biow0lf** [comment]
@jachenry any news?

**Comment by mvastola** [comment]
Seeing now that other people ran have since run into this bug (see #45). Perhaps one or more of these people can test this fix and confirm it works in lieu of @jachenry so we can get this pushed ASAP, considering `rails` was just released today?

I'll go ahead and fix the merge conflicts on this.

CC: @cannikin, @ZempTime, @bodrovis

**Comment by ericboehs** [comment]
Looks like Travis is failing.

> rails/rails-observers/lib/rails/observers/active_model/observing.rb:19:in `inherited': uninitialized constant ActiveModel::Observing::ActiveRecord (NameError)

@mvastola could you take a look

File: lib/rails/observers/active_model/observing.rb [hunk 1]
```diff
@@ -14,8 +14,13 @@ module Observing
 
     included do
       extend ActiveSupport::DescendantsTracker
+      def self.inherited(subclass)
+        super
+        ActiveRecord::Base.instantiate_observers
```
Comment (CloCkWeRX) [inline lib/rails/observers/active_model/observing.rb]: Since this is in ActiveModel, not activerecord; should it be self.instantiate_observers?

Also in the railtie; there's a lot of loading and explicitly instantiate_observing going on; but nothing that specifically seems to target active_model itself

Comment (kwent) [inline lib/rails/observers/active_model/observing.rb]: The issue is 

```
/home/travis/build/rails/rails-observers/lib/rails/observers/active_model/observing.rb:19:in `inherited': uninitialized constant ActiveModel::Observing::ActiveRecord (NameError)
```

What about escaping the namespace ?

```
::ActiveRecord::Base.instantiate_observers
```

**Comment by panigrah** [comment]
@mvastola CI build fails  - are you working to resolve the circular dependency issue in the test now?

**Comment by jnimety** [comment]
Any updates on this issue?

**Comment by rafaelfranca** [comment]
No updated, as you can see in the conversation in this PR.

**Comment by jnimety** [comment]
is the short term work around to add an initializer with `ActiveRecord::Base.instantiate_observers`?

File: lib/rails/observers/active_model/observing.rb (hunk 1 above)
Comment (enrico) [inline lib/rails/observers/active_model/observing.rb]: @mvastola , @rafaelfranca : Just took a look at this and and the Travis tests fail because in the context of the active_resource_observer_test, ActiveRecord::Base is not defined at all and therefore not found. 
I'm not clear whether the gem can be used without activerecord or not. I'm also not sure if we can use Object.const_defined?('ActiveRecord::Base') to verify existence of ActiveRecord::Base...

Either way, this fixes all tests, you guys might have a more elegant solution to it...:

```
ActiveRecord::Base.instantiate_observers if Object.const_defined?('ActiveRecord::Base')
```

**Comment by andela-kadeniyi** [comment]
@mvastola Can I take a stab at getting to see why this spec fails?

**Comment by mvastola** [comment]
@andela-kadeniyi I'm so sorry. I had already [intimated](https://github.com/rails/rails-observers/pull/39#issuecomment-242992661) that the task of fixing this was up for the taking, but I'm now realizing I did so only in a comment on the (merged/closed) PR #39 whose regression this PR fixes. 

For convenience, I'm duplicating [my comments](https://github.com/rails/rails-observers/pull/39#issuecomment-242992661) below.

> Hey guys,
> The real issue here is in #41, so it would probably be best if all comments were redirected there. It would be even better if someone wanted to help work on that PR as my time is limited. (I'm happy to grant access to my fork.)
> 
> Long story short:
> I introduced a regression in this PR (#39) which, after it was pointed out to me, I promptly fixed with #41 (which, at the time, passed all CI tests and was merge-conflict-free).
> 
> Unfortunately, by the time someone else verified the fix in #41, commits had already been made that resulted in merge conflicts between #41 and HEAD. While these conflicts were simple enough to resolve (and have been resolved in #41 as it stands), the CI tests no longer pass (suggesting the commits to master after #39 was merged made some incompatible changes) and thus far, I have not been able to successfully resolve the errors.

Basically, as you can see above, this PR worked (for me) at first, but due to a real-life race condition, I couldn't get someone to verify the PR prior to some conflicting commits being made. 

Once they were, I resolved the merge conflicts, but for some reason (which, if I understood, I could resolve this) the commits that were made conflict with my PR on a functional level and cause Travis to fail. What you see now is as far as I've gotten. Personal time constraints prevent me from giving this more time at the moment, but I'm happy to assist in any way I can. Feel free to give me a prod over email if I don't reply for a few days.

**Comment by enrico** [comment]
@mvastola , did you see my comment above on why the test fails?  (Travis tests fail because in the context of the active_resource_observer_test, ActiveRecord::Base is not defined and therefore not found) Hopefully this puts you in the right direction...

**Comment by mvastola** [comment]
@andela-kadeniyi, actually I think I might have this. Can you hold off a few hours?
Thanks!

**Comment by mvastola** [comment]
@enrico, I hadn't. I actually just saw it (right after I sent my previous reply). I'm testing it out, but it seems to be working. (I suspect I'll be thanking you a ton in a few.)

`ActiveRecord` is -- indeed -- a dependency, so I think the problem may have been that it wasn't initialized by the CI test, meaning my patch was correct and the test was wrong and this was a wild goose chase.  🤦‍♂️

**Comment by mvastola** [comment]
Alright, I'm going to get this thanks to @enrico. Hoping to have it done tomorrow.

**Comment by mvastola** [comment]
Still working on this btw. Sorting out a few things still.

**Comment by mvastola** [comment]
So, update. I put a lot more hours into this, and I'm back in circular dependency hell again. Now I'm getting annoyed, which means I'm determined to solve this. 

Depending on what I do, the issue seems to be going back and forth between this circular dependency thing and just a silent failure to instantiate the observers. (Though I think I have a good handle on avoiding the latter issue.)

I spent a bit of time neatening up the code in the hopes that would -- if not fix things in and of itself -- make the solution more straightforward. (@enrico, I honestly thought your solution *had* been working entirely, but for some reason it doesn't get past everything now, even if apply the exact change you mentioned to the exact code you initially commented on. Any thoughts? One thing to note is I had to tweak your code slightly because I noticed that that module isn't just included in `ActiveRecord::Base`, but also `ActiceResource::Base`, so I added a method to look at the superclasses of the observed class and call `#instantiate_observers` on the closest ancestor that had `include`d `ActiveModel::Observing`, which seems to get past that issue.) 

Basically, what seems to be happening is:
1. The first `ActiveRecord` model to be referenced causes that model's file -- say `app/models/user.rb` -- to be autoloaded, as it is expected to define the class).
1. As a result `ActiveRecord:::Base` is loaded because it is the superclass.
1. This triggers the `ActiveSupport::LazyLoadHooks` we have set up in our `Railtie` to run as soon as `ActiveRecord::Base` is loaded.
1. These hooks require the `rails-observer` magic, extend `ActiveRecord:::Base`, and instantiate all the `ActiveModel::Observer`s referred to in `config/application.rb`.
1. When an `ActiveModel::Observer` is instantiated, `rails-observers` populates an array, stored in an instance variable with the actual classes of the models it observes. 
1. By constantizing the names of these models in the previous step, an attempt is once again made to load, say `app/models/user.rb`. Creating our issue. (Previous versions of `ActiveSupport` apparently have different circular dependency checks, making this issue unique to Rails 5.)

What I'm planning to do, (and which I have every reason to think will work) is to try to switch over `rails-observers` to *not* autoload observed models and instead hook into them lazily if and when they are loaded naturally via `self.inherited` callback in `ActiveRecord::Base`. I can't imagine this should cause any issues since, if the model hasn't loaded yet, by definition, there is nothing to observe.

The alternative solution (which should also work) should be to eager load all listed `ActiveModel::Observer`s in the `Railtie`, but that doesn't seem to be as good an idea because that implicitly eager loads `ActiveRecord::Base`, all *it's* dependencies, and possibly instantiates database connections as well.

Let me know if anyone has any thoughts, but that's the plan for now.

**Comment by domcleal** [comment]
> What I'm planning to do, (and which I have every reason to think will work) is to try to switch over rails-observers to not autoload observed models [..]

Reading 26d37c5, I'm unsure how you intend to do this. It looks like you plan to disable autoloading while the observer is being loaded, but then constants used in `observe Foo` will be undefined. That'd also mean autoloading for other things, such as parent classes and concerns wouldn't be possible.

As long as constants are supported in the observer's `observe` class statement, I think your alternative solution is preferable:

> The alternative solution (which should also work) should be to eager load all listed ActiveModel::Observers in the Railtie [..]

The dependency order is much clearer this way: the app loads the observer, the observer loads the model and AR::Base, and then the model has no (circular) dependency back on the observer.

In the longer term I think rails-observers could/should switch to loading observers only on `self.inherited`, and as you say, this will load less code and avoids registering observers for models that aren't even used yet. This may require removing support for constants in the `observe` line and requiring that users specify the class by symbol or string, like AR's `has_*` associations use `:class_name`. This would prevent any autoload dependency from the observer onto the model, so observers could be safely loaded with the model via `self.inherited`.

In the meantime, I think the alternative solution is more simple and reliable than disabling autoloading, and then support for constants can be deprecated so it can be refactored in a future major version.

**Comment by mvastola** [comment]
@domcleal, turning off autoloading was something I was only playing with. I removed that. (Code is getting there, btw.) 

I was basically going to do what you said: enabling support for strings/symbols in the observe method and watch for class creation in `self.inherited`. I was going to convert any classes into strings for backwards compatibility, but you're making me realize that won't always work. As long as the autodetected class name is also made a string though, the (immediate) circular dependency should be removed as most user probably haven't manually specified `observe` for the implied model. (I guess it's something to check for if a larger circular dependency loop can be made.) 

I'm not sure eager loading is a viable option. Some people have huge apps and in development mode reloading needs to be done on a per-request basis. (There's also a definite benefit to doing it lazily there because it allows for the testing of different dependency load orders.) And considering this change is due to a major Rails version upgrade, I think it's appropriate if we took this time to move to stringy class names. 

Open to feedback though.

**Comment by mvastola** [comment]
PS: Or maybe there's some way with `instance_exec` and a `Module` with a prepended `const_missing` to create some syntax sugar so that we can convert the constants to strings?

**Comment by domcleal** [comment]
Yep, I think using strings/symbols throughout for the observed class names would probably be the ideal solution.

Fair point about it being for a major Rails upgrade though, it's reasonable to change the format of this. The README does show the use of constants in the `observe` line, but that's easily updated.

If accepted, then issuing a 0.1.2 release with a deprecation warning might be sufficient to help users change definitions rather than continuing to handle constants in `master` or 1.x (by `const_missing` or otherwise.)

**Comment by kgeoir** [comment]
Hello, any news about this issue?
I having my oberservers working fine in dev env but not at all in production....

**Comment by mvastola** [comment]
@kgeoir, yes. My crazy life has just been getting in the way. I made a bunch of changes while trying to find out the best way to implement this, many of which (however useful) ended up being irrelevant (and thus will be beyond the scope of this PR; expect other PRs once this is done). Several other conflicting projects have also been taking up my time. 

I really ought to at least sort out everything I've done so far into discrete commits (expect that today or tmrw) and I truly am in a position to get this done very soon. I'm thinking Sunday.

Question though for @dhh, I have this repo on my RSS reader and you've been making some commits the past couple of days with comments that seem to be suggesting you're gearing up to release Rails 5.1 compatibility. Is that an allusion to this PR, or is someone else working on something instead because I've been admittedly lagging on this?

**Comment by dhh** [comment]
Mike, I was just trying to get this gem released for 5.1 compatibility. But
if there are some concerns you need to address before that happens, please
do!

**Comment by mvastola** [comment]
@dhh, this PR is working on 5.1 compatibility. Does that mean you're working on it as well? How far have you gotten? Maybe we should collaborate? 

I don't have any concerns, I just wanted to know if you were about to swoop in and supersede my work on this, in which case I wouldn't bother continuing. If you have time to spend on this though, it might be a good idea to work together.

**Comment by mvastola** [comment]
Er. "Swoop in" came out the wrong way. To be clear, I wouldn't begrudge you one bit. (This has taken far too long, and 5.2 is now out, which should also be supported). I just noticed that someone was making commits about this and hadn't said anything in the primary related PR.

**Comment by dhh** [comment]
I haven't done any actual work on the gem for compatibility. I only nixed
the upper-bound. So please do proceed with any actual compatibility work.
Thanks!

**Comment by mvastola** [comment]
Okay, so this doesn't work yet, but lest people think nothing has been done on this, I wanted to push something. 

(And, I know, I broke a major golden rule of coding by fixing about a gazillion things in these commits -- to the point that this is pretty much a total refactor -- so I anticipate this will be parsed out into a number of smaller PRs, but I saw a number of unintuitive things with the code organization at first and I couldn't be sure that they weren't the problem, and things just escalated from there.)

Things that are welcome:
* All constructive criticism
* General comments
* Offers to do any part of work that remains, including but not limited to
  * Adapting existing unit tests
  * Writing new unit tests for additional functionality/features/facets of the code
  * Simply pulling out a distinct function/feature/facet of the code into a separate PR for merging.

I'm sure someone industrious can probably also take about 20 lines from all the code I wrote and the steps I laid out and fix this one major bug, which would also be fine and simultaneously put me to shame. :-p

**Comment by kaspth** [comment]
Why can't we just add back the `instantiate_observers` line here? https://github.com/rails/rails-observers/commit/bca5db98e2e81d3dbf5fe001e55781951e248f3c#diff-9db46e46113d0c2eeb41f5d6f08bcac4L24

**Comment by mvastola** [comment]
@kaspth Heh. Because then you get [this](https://github.com/Partyista/rails-observers/blob/aeca1124b5d7bd02fcc2b032895b5544ce134c52/NOTES-MV.md#stack-trace).

**Comment by kgeoir** [comment]
I am not sure to understand all the issue. As per my quick investigation it seems form me that the code in to_prepare is never call. It seems for me like an issue with the reloader. I tested it with Rails 5.0.2 and 5.0.3.

Meanwhile what I did as a dirty fix so I can move forward will the good fixe is merged, is I added 
`    config.after_initialize do |app|
      ActiveSupport.on_load(:active_record) do
        ActiveRecord::Base.instantiate_observers
      end
    end`
in my application.rb. it seems ok for now.

```

      config.after_initialize do |app|
        ActiveSupport.on_load(:active_record) do
          # Rails 5.1 forward-compat. AD::R is deprecated to AS::R in Rails 5.
          reloader = defined?(ActiveSupport::Reloader) ? ActiveSupport::Reloader : ActionDispatch::\
Reloader
          reloader.to_prepare do
            ActiveRecord::Base.instantiate_observers
          end
        end
```

**Comment by mvastola** [comment]
Have you read these comments in particular? [this](https://github.com/rails/rails-observers/pull/41#issuecomment-276468791) and [this](https://github.com/rails/rails-observers/pull/41#issuecomment-277473341)

AFAIK, this problem isn't occurring in dev mode. Only in production. So reloaders don't apply. Although there could be an extra bug in reloading (or, I would imagine, those reloaders don't apply if the code wasn't loaded in the first place).

My understanding of the issue is/was that the issue is caused by eager loading in production.

Basically if the observers are instantiated at the correct time given the code as it stands (which they need to be in order to properly track changes that could be made by rails initializers and properly-configured ActiveSupport.on_load hooks) as of Rails 5, they will actually cause the load of the model class they observe, which then causes a circular dependency because the Observer immediately hooks in while the class is loading and then tries to reference the class that hasn't fully loaded. (See stack trace.)

**Comment by mvastola** [comment]
FWIW, I am aware that a quick and dirty workaround does exist involving instantiating the observers. The problem though is unless the observers hook in immediately upon loading, the observations really aren't continuous.

Also, I should note, all this discussion has been about AR, but theoretically `rails-observers` is ORM agnostic and can observe `ActiveModel`, `ActiveResource` or technically any old class that just does an `include ActiveModel::Observing`. Point though is that it's possible that this whole incomplete coverage/regression/circular dependency problem applies to that code too and those are just less common cases or they haven't brought it up here because it doesn't seem to be a problem with *this* code.

**Comment by kaspth** [comment]
I've committed a fix for the "circular reference" root issue over here: https://github.com/rails/rails-observers/pull/65

Appreciate the deep investigation and help with supporting newer Rails releases, @mvastola! ❤️

**Comment by mvastola** [comment]
@kaspth, you're my hero!!

Thanks so much for boiling this down into a simple fix for now -- I obviously hadn't had as much time as I'd expected to bring this PR into a working state. 

So obviously this PR suffered from abominable levels of feature creep because I had no idea what would fix the problem (I thought it would have been more tedious to go through all the commits to rails and AR -- your approach -- but hindsight is 20/20). 

In the end though, I implemented -- or mostly implemented -- many of the features I sought to (just all jumbled together). I'd hate to see this work and code go to waste though. 

Is there any way you could take a minute and let me know your thoughts on the various features I began and particularly their suitability for incorporation (assuming they're submitted in their own PRs with appropriate tests and all that jazz)? 

The features I'm referring to include: 
* Regorganizing the directory structure and some module/class names
* Removed dependency on rails/railties (while adding a compatibility wrapper)
* Allow for lazy loading of observers, models they observe, and the `ActiveRecord::Base` class (by referencing observed by classes by symbol/string)

**Comment by kaspth** [comment]
> I thought it would have been more tedious to go through all the commits to rails and AR -- your approach

Hardly inspected any commits. I reverted #39 to add Rails versions back manually. Rails 4.1 started exhibiting the circular reference bug. I slipped in a byebug statement around `instantiate_observers` and noticed the different backtraces. In Rails 4.0 the call came from before the `user:count` Rake task ran, while 4.1 had it from running `class User < ActiveRecord::Base`.

> Regorganizing the directory structure and some module/class names

I don't care enough to review that 😊

> Removed dependency on rails/railties (while adding a compatibility wrapper)

Not needed. Everything is wrapped in a railtie that's only loaded when `Rails::Railtie` is defined.

> Allow for lazy loading of observers, models they observe, and the ActiveRecord::Base class (by referencing observed by classes by symbol/string)

Could be good. It would match Active Record associations' `class_name: "User"` option.

**Comment by mvastola** [comment]
> Hardly inspected any commits. I reverted #39 to add Rails versions back manually

I guessed. I would have done that first if I knew what this would become. :-)

> I don't care enough to review that 😊

There's an ancient PR (#13) regarding this. Maybe I'll try to do it by writing a script that does the re-org so that's the only thing that needs reviewing.

> Not needed. Everything is wrapped in a railtie that's only loaded when `Rails::Railtie` is defined.

See what I did [here](https://github.com/Partyista/rails-observers/blob/6b18e75022fdc3696ed5aef056ccac443fdc4e76/lib/rails/observers/boot.rb) and [here](https://github.com/Partyista/rails-observers/blob/6b18e75022fdc3696ed5aef056ccac443fdc4e76/lib/rails/observers/bootstrapper.rb). It's not done very robustly, and IMHO this should be part of AS, but the idea is to have an abstraction layer that can bootstrap the `AS::LazyLoadHook`s and the `Reloader` (i.e. everything rails-observers uses, without the rails dependency).

**Comment by kaspth** [comment]
I'll take a pass on those, thanks 😊
